### PR TITLE
[TIMOB-26484] Android: Added "sound" support to NotificationChannel

### PR DESCRIPTION
**Pull request by jquick-axway**
**JIRA:**
https://jira.appcelerator.org/browse/TIMOB-26484

**Note:**
This is needed to resolve issue [TIMOB-26432](https://jira.appcelerator.org/browse/TIMOB-26432), where custom sounds attached to individual notifications are ignored on Android 8.0 and newer OS versions. For these OS versions, you must assign the custom sound to a notification channel instead.

**Test:**
1. Create a Titanium classic project with `NotificationSoundTest.js` attached to [TIMOB-26484](https://jira.appcelerator.org/browse/TIMOB-26484).
2. Under your project, create folder: `./platform/android/res/raw`
3. Copy `notification_sound.wav` to the above raw folder.
4. Play this sound file on your desktop so that you'll know what it sounds like later.
5. Build and run this project on an Android 8.0 or newer device.
6. Tap the "Send Notification" button.
7. Verify that you hear the above `notification_sound.wav`. _(This is the fix.)_
8. In the log, verify that you see warning message: `Notification 'sound' property is not supported on Android 8.0 and higher. You must assign sound to its NotificationChannel instead.`
9. Run the same project on an Android 7.1 or older device.
10. Tap the "Send Notification" button.
11. Verify that you hear the `notification_sound.wav`. _(Verifies old method still works.)_

### CODE REVIEW

File: apidoc/Titanium/Android/NotificationChannel.yml [hunk 1]
```diff
@@ -78,6 +78,20 @@ properties:
...
+        The URL cannot refer to a file in the Resources directory, but it can refer to
+        a file in the raw resources directory
+        ([Filesystem.resRawDirectory](Titanium.Filesystem.resRawDirectory)).
+
+        To create a raw resource, place the audio file in the `platform/android/res/raw`
+        folder and reference it as:
+
+            notificationChannel.sound = Ti.Filesystem.resRawDirectory + 'my_sound.wav';
+
+    type: String
```
Comment (ypbnv) [inline apidoc/Titanium/Android/NotificationChannel.yml]: Maybe add a since property here?

**Review by ypbnv — APPROVED** [review]
Looks good to me!

**Review by garymathews — APPROVED** [review]
CR: See comments

File: android/modules/android/src/java/ti/modules/titanium/android/notificationmanager/NotificationProxy.java [hunk 2]
```diff
@@ -384,10 +384,16 @@ public void setNumber(int number)
 	public void setSound(String url)
 	// clang-format on
 	{
+		if (Build.VERSION.SDK_INT >= 26) {
+			Log.w(TAG, "Notification 'sound' property is not supported on Android 8.0 and higher. "
+						   + "You must assign sound to its NotificationChannel instead.");
+		}
```
Comment (garymathews) [inline android/modules/android/src/java/ti/modules/titanium/android/notificationmanager/NotificationProxy.java]: Shouldn't this return?

File: apidoc/Titanium/Android/NotificationChannel.yml (hunk 1 above)
Comment (jquick-axway) [reply apidoc/Titanium/Android/NotificationChannel.yml]: You are right. I'll add it.

File: android/modules/android/src/java/ti/modules/titanium/android/notificationmanager/NotificationProxy.java (hunk 2 above)
Comment (jquick-axway) [reply android/modules/android/src/java/ti/modules/titanium/android/notificationmanager/NotificationProxy.java]: Google's API no-ops on Android 8.0+. So, I didn't see any harm in continuing.

What I'm not sure if I like is logging this every time `Notification.sound` is set. It makes sense to log it to **inform** the app developer that they need to use it for Android 8.0+. But at the same time, they need to set this property for Android 7.1 and below anyways and it's convenient to set this property without an API Level check in the code... making this warning message annoying.

**Comment by lokeshchdhry** [comment]
@jquick-axway , Local notifications are fine. But, I am having issues with GCM push. I am getting this error with an SDK with this fix. 7.4.1.GA is fine :
```
[ERROR] :  [Nexus 6P] TiExceptionHandler: (main) [1027,1027]
[ERROR] :  [Nexus 6P] TiExceptionHandler:     com.appcelerator.aps.CloudpushModuleImplementation$10.getMobileId(CloudpushModuleImplementation.java:251)
[ERROR] :  [Nexus 6P] TiExceptionHandler:     com.appcelerator.aps.CCPushService.injectAnalytics(CCPushService.java:124)
[ERROR] :  [Nexus 6P] TiExceptionHandler:     com.appcelerator.aps.CCPushService.getGCMSenderId(CCPushService.java:155)
[ERROR] :  [Nexus 6P] TiExceptionHandler:     com.appcelerator.aps.CCPushService.access$000(CCPushService.java:35)
[ERROR] :  [Nexus 6P] TiExceptionHandler:     com.appcelerator.aps.CCPushService$1.doInBackground(CCPushService.java:97)
[ERROR] :  [Nexus 6P] TiExceptionHandler:     com.appcelerator.aps.CCPushService$1.doInBackground(CCPushService.java:89)
[ERROR] :  [Nexus 6P] TiExceptionHandler:     android.os.AsyncTask$2.call(AsyncTask.java:333)
[ERROR] :  [Nexus 6P] TiExceptionHandler:     java.util.concurrent.FutureTask.run(FutureTask.java:266)
[ERROR] :  [Nexus 6P] TiExceptionHandler:     android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:245)
[ERROR] :  [Nexus 6P] TiExceptionHandler:     java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162)
[ERROR] :  [Nexus 6P] TiExceptionHandler:     java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636)
[ERROR] :  [Nexus 6P] TiExceptionHandler:     java.lang.Thread.run(Thread.java:764)
```

I am using this code:
```
var soundPath = Ti.Filesystem.resRawDirectory + "notification_sound.wav";

var Cloud = require('ti.cloud'),
	CloudPush = require('ti.cloudpush'),
	PlayServices = require('ti.playservices'),

	username = 'test12345',
	password = 'test12345',
	type = Ti.Platform.name === 'iPhone OS' ? 'ios' : Ti.Platform.name,
	channelId = 'test_channel123',
	token = null,
	user = null,

	win = Ti.UI.createWindow({ backgroundColor: 'gray' }),
	lbl = Ti.UI.createLabel({
		color: 'white',
		top: 5,
		left: 5,
		text: 'logging in...\n'
	});
	channel = Ti.Android.NotificationManager.createNotificationChannel({
        id: channelId,
        name: 'TEST CHANNEL',
        importance: Ti.Android.IMPORTANCE_DEFAULT,
        sound: soundPath,
    });

Ti.Android.requestPermissions('android.permission.WRITE_EXTERNAL_STORAGE', function(e){
	if (e.success) {
		win.add(lbl);
		win.open();
		Cloud.Users.login({
			login: username,
			password: password
		}, function(e) {
			if (e.success) {
				var user = e.users[0];
				lbl.text += 'logged in as ' + user.id + '\n';
				subscribe();
			} else {

				// create account
				if (e.code === 401) {
					Cloud.Users.create({
						username: username,
						password: password,
						password_confirmation: password,
						first_name: username,
						last_name: username,
						email: username + '@email.com'
					}, function(e) {
						if (e.success) {
							user = e.users[0];
							lbl.text += 'created user ' + user.id + '\n';
							subscribe();
						} else {
							lbl.text += 'error creating user... (see logcat)\n';
							Ti.API.error(JSON.stringify(e, null, 2));
						}
					});
				} else {
					lbl.text += 'error logging in... (see logcat)\n';
					Ti.API.error(JSON.stringify(e, null, 2));
				}
			}
		});
	} else {
		alert('could not acquire storage permissions');
	}
});

function subscribe () {
	if (PlayServices.isGooglePlayServicesAvailable() === PlayServices.RESULT_SUCCESS) {
		CloudPush.retrieveDeviceToken({
			success: function(e) {
				token = e.deviceToken;
				lbl.text += 'token: ' + token + '\n';

				Cloud.PushNotifications.subscribe({
					channel: channelId,
					device_token: token,
					type: type
				}, function(e) {
					if (e.success) {
						lbl.text += 'subscribed to channel: ' + channelId + '\n';

						function callback(e) {
							alert(JSON.stringify(e, null, 2));
						}

						CloudPush.addEventListener('callback', callback);
						CloudPush.addEventListener('trayClickFocusedApp', callback);
						CloudPush.addEventListener('trayClickLaunchedApp', callback);

						lbl.text += 'waiting for notifications...';
					} else {
						lbl.text += 'error subscribing... (see logcat)\n';
						Ti.API.error(JSON.stringify(e, null, 2));
					}
				});
			},
			error: function(e) {
				lbl.text += 'error retreiving token... (see logcat)\n';
				Ti.API.error(JSON.stringify(e, null, 2));
			}
		});
	} else {
		alert('Google Play Services is not installed/updated/available');
	}
}
```

**Comment by garymathews** [comment]
@lokeshchdhry I've updated the `cloudpush` module, test this PR with [cloudpush-5.2.0](https://github.com/appcelerator-modules/cloudpush/releases/tag/android-5.2.0).

**Comment by lokeshchdhry** [comment]
@garymathews , Cloudpush 5.2.0 throws:
```
[ERROR] :  Conflicting jar files detected:
[ERROR] :  
[ERROR] :  The following modules have different "cloudpush.jar" files
[ERROR] :     ti.cloudpush   (version 5.2.0) (hash=5cf89ff6dcdd18c6be95e7d9098fa2cd)
[ERROR] :     ti.cloudpush   (version 5.2.0) (hash=818fe6b00dcd4fa6f53bbeae31a63504)
[ERROR] :  
[ERROR] :  You can either select a version of these modules where the conflicting jar file is the same or you
[ERROR] :  can try copying the jar file from one module's "lib" folder to the other module's "lib" folder.
[ERROR] Application Installer abnormal process termination. Process exit value was 1

```

**Comment by lokeshchdhry** [comment]
@garymathews , @jquick-axway , I have moved the CP issue to a separate ticket : https://jira.appcelerator.org/browse/TIMOB-26491

**Comment by lokeshchdhry** [comment]
@jquick-axway , I will merge this PR as local notifications are working & it does not depend on the Cloudpush issue. 
In case we have issues later will follow it up with separate ticket.

**Comment by lokeshchdhry** [comment]
FR Passed.

Local notifications seem to work as expected on Android 8.0+ & below.
GCM push test will be checked once https://jira.appcelerator.org/browse/TIMOB-26491 is fixed & if issues found will be addressed in a separate ticket.

Studio Ver:     5.1.2.201810080801
SDK Ver:        7.5.0 local build
OS Ver:         10.14
Xcode Ver:      Xcode 10.0
Appc NPM:       4.2.13
Appc CLI:       7.0.6
Daemon Ver:     1.1.3
Ti CLI Ver:     5.1.1
Alloy Ver:      1.13.2
Node Ver:       8.9.1
NPM Ver:        5.5.1
Java Ver:       10.0.2
Devices:        ⇨ google Nexus 5 (Android 6.0.1)
                ⇨ google Nexus 6P (Android 8.1.0)

**Comment by lokeshchdhry** [comment]
FR Passed.

Notification sounds works as expected on Android 8.0+ & below as well.

NOTE: Due to googles changes you will have to add sound to individual notification for Android 7.1 and below. While for 8.0+ you will have to add sound to the notification channel.

Also, the `channelId` in the notification channel for Android 8.0+ should be in the `android` table of the JSON payload like below:
```
{
  "android": {
    "title": "High score ",
    "alert": "You have a new score",
    "channelId": "test_channel1234567"
  }
}
```

Studio Ver:     5.1.2.201810080801
SDK Ver:        7.5.0 local build
OS Ver:         10.14
Xcode Ver:      Xcode 10.0
Appc NPM:       4.2.13
Appc CLI:       7.0.6
Daemon Ver:     1.1.3
Ti CLI Ver:     5.1.1
Alloy Ver:      1.13.2
Node Ver:       8.9.1
NPM Ver:        5.5.1
Java Ver:       10.0.2
Devices:        ⇨ google Nexus 5 (Android 6.0.1)
                ⇨ google Nexus 6P (Android 8.1.0)

**Comment by tbao1** [comment]
@lokeshchdhry  I am running into problems including the channelID into the payload, when we define the Android dictionary, we end up not being able to send iOS notifications, is there anyway we can have both dictionarys working at the same time?

**Comment by Max87ZA** [comment]
> @lokeshchdhry I am running into problems including the channelID into the payload, when we define the Android dictionary, we end up not being able to send iOS notifications, is there anyway we can have both dictionarys working at the same time?

Same problem here, is there a solution to send PN to both platforms with custom sound?

**Pull request merged**